### PR TITLE
sstable: fix indexBlockBuf.shouldFlush by accounting for inflight data

### PR DESCRIPTION
**Pull request by bananabrick**
When we compute the number of entries in indexBlockBuf.shouldFlush, we need
to account for the number of inflight entries. Without accounting for inflight
entries, it is possible for the index block sizes to balloon.

### CODE REVIEW

**Review by bananabrick — COMMENTED** [review]
This pr is based off of: https://github.com/cockroachdb/pebble/pull/1544. I had to make some ordering changes/rebase the commits in #1544, but the pr tiny enough that re-writing it was more convenient. If that's an inconvenience, I can go back to that pr.

Flushing some responses to comments in #1544.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581#-:-My9NfobDdijNInsNL5s:b-7mqrdc)* status: 0 of 7 files reviewed, 2 unresolved discussions (waiting on @itsbilal and @jbowens)

___
*[options.go, line 487 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1581#-My9PHYm1Q5oC2egdSFW:-My9PHYm1Q5oC2egdSFX:b-j77zvu) ([raw file](https://github.com/cockroachdb/pebble/blob/ae80cfc8a413917bc3615afea9645f8e3e45905d/options.go#L487)):*
> ```Go
> 		// compress data blocks and write datablocks to disk in parallel with the Writer client
> 		// goroutine.
> 		EnableWriterParallelism bool
> ```

From @itsbilal 

```
nit: This could just be WriterParallelism or ParallelCompression, the Enable is implied with the bool. Though I'd prefer if this were an int that specified the max number of threads we'd create per writer, defaulting to 1 (no parallelism).,
```

Done
___
*[sstable/writer.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1581#-My9PkCRCHTeghDTv5wA:-My9PkCRCHTeghDTv5wB:bf45lqv) ([raw file](https://github.com/cockroachdb/pebble/blob/ae80cfc8a413917bc3615afea9645f8e3e45905d/sstable/writer.go#L225)):*
> ```Go
> 	// By default, we use a writeQueue size of 0, since we won't be doing any block writes in parallel.
> 	var writeQueueSize int
> 	if parallelismEnabled {
> ```

The primary comment in #1544 was:

by @itsbilal 
```
I think it'd make sense to have WriterParallelism be an int option which defaults to 1 (no parallelism) and if increased beyond 1, is the max number of threads we'll spawn per writer, going up to min(that option, GOMAXPROCS(0)). It makes sense to give the spawner of pebble some control over the max value of this parallelism, just like we do with compaction concurrency and other concurrency gates. We will likely have multiple writers open at once, and it seems inefficient for us to go straight to letting them spawn as many threads as the number of cores. Cockroach could, for example, default to a value of GOMAXPROCS(0)/4, or a constant like 2-4.
```

by @nicktrav 
```
Using an int has the added benefit that the metamorphic test generator can set these numbers randomly (within some interval), and it may help catch issues that only manifest with a high degree of parallelism / concurrency.
```

There's a few reasons why I don't want to make the write queue size a configurable parameter, but I'm open to a discussion.

1. The write queue size doesn't determine the number of goroutines spawned by the `writeQueue`. It's just the maximum number of elements which can be added to the writeQueue, before blocking. It does determine the amount of disk io done, though.

2. The goal is to write a system with fewer tuneable parameters, and where the system is smart enough at detecting cpu, that the operator won't have any problems. If the operator does notice high cpu, then the feature can be completely disabled. MaxCompactionConcurrency is a good example, because it's a parameter we're trying to get rid of in the future by auto-tuning it.

3. Setting the write queue size to `runtime.GOMAXPROCS(0)`, is currently arbitrary. I wanted to set it to a value greater than 0, to enable parallelism in the metamorphic tests. Allowing the metamorphic tests to control the size of the queue isn't useful, because they're just controlling the number of elements added to the queue before it blocks. There is no additional concurrency, or edge cases being tested.

4. In a future pr, I plan to set the writeQueue size such that the amount of disk IO done is comparable to the current master.

**Comment by bananabrick** [comment]
Adding tests

**Comment by bananabrick** [comment]
Should be ready for review.

**Review by mufeez-amjad — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 3 of 6 files at r3, 2 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581)* status: 6 of 8 files reviewed, 5 unresolved discussions (waiting on @bananabrick, @itsbilal, and @jbowens)

___
*[sstable/options.go, line 222 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyTj6x50bfXudhY-D3j:-MyTj6x50bfXudhY-D3k:bf5jvzo) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/sstable/options.go#L222)):*
> ```Go
> 	// Parallelism is used to indicate that the sstable Writer is allowed to compress
> 	// data blocks and write datablocks to disk in parallel with the Writer client goroutine.
> 	// goroutine.
> ```

nit: extra word
___
*[sstable/writer.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyTlfw-C6iX0TA27JHF:-MyTlfw-C6iX0TA27JHG:b-fsr28q) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/sstable/writer.go#L224)):*
> ```Go
> 
> 	// By default, we use a writeQueue size of 0, since we won't be doing any block writes in parallel.
> 	var writeQueueSize int
> ```

nit: I think `writeQueueSize := 0` would be more readable.
___
*[sstable/writer.go, line 361 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyTkTYh2RvLmUPDtmwS:-MyTkTYiDcEJICzNWMtJ:b-7k4ybb) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/sstable/writer.go#L361)):*
> ```Go
> 	}
> 
> 	nEntries := i.size.estimate.numWrittenEntries + i.size.estimate.numInflightEntries
> ```

nit: consider making this a function on `sizeEstimate`

**Review by itsbilal — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r2, 1 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581)* status: all files reviewed, 7 unresolved discussions (waiting on @bananabrick, @jbowens, and @nicktrav)

___
*[options.go, line 911 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyhHyEB94yYjZcs-NSi:-MyhHyEB94yYjZcs-NSj:b-grrslm) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/options.go#L911)):*
> ```Go
> 	fmt.Fprintf(&buf, "  wal_dir=%s\n", o.WALDir)
> 	fmt.Fprintf(&buf, "  wal_bytes_per_sync=%d\n", o.WALBytesPerSync)
> 	fmt.Fprintf(&buf, "  enable_writer_parallelism=%t\n", o.Experimental.WriterParallelism)
> ```

This could also just be `writer_parallelism`
___
*[options.go, line 1129 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyhI0VI3FPY9d83R_cj:-MyhI0VI3FPY9d83R_ck:bb73xfv) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/options.go#L1129)):*
> ```Go
> 				o.WALBytesPerSync, err = strconv.Atoi(value)
> 			case "enable_writer_parallelism":
> 				o.Experimental.WriterParallelism, err = strconv.ParseBool(value)
> ```

Same
___
*[sstable/writer.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1581#-My9PkCRCHTeghDTv5wA:-MyhIAWS3flNndMz_OqA:bi1lhpx) ([raw file](https://github.com/cockroachdb/pebble/blob/ae80cfc8a413917bc3615afea9645f8e3e45905d/sstable/writer.go#L225)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

The primary comment in #1544 was:

by @itsbilal 
```
I think it'd make sense to have WriterParallelism be an int option which defaults to 1 (no parallelism) and if increased beyond 1, is the max number of threads we'll spawn per writer, going up to min(that option, GOMAXPROCS(0)). It makes sense to give the spawner of pebble some control over the max value of this parallelism, just like we do with compaction concurrency and other concurrency gates. We will likely have multiple writers open at once, and it seems inefficient for us to go straight to letting them spawn as many threads as the number of cores. Cockroach could, for example, default to a value of GOMAXPROCS(0)/4, or a constant like 2-4.
```

by @nicktrav 
```
Using an int has the added benefit that the metamorphic test generator can set these numbers randomly (within some interval), and it may help catch issues that only manifest with a high degree of parallelism / concurrency.
```

There's a few reasons why I don't want to make the write queue size a configurable parameter, but I'm open to a discussion.

1. The write queue size doesn't determine the number of goroutines spawned by the `writeQueue`. It's just the maximum number of elements which can be added to the writeQueue, before blocking. It does determine the amount of disk io done, though.

2. The goal is to write a system with fewer tuneable parameters, and where the system is smart enough at detecting cpu, that the operator won't have any problems. If the operator does notice high cpu, then the feature can be completely disabled. MaxCompactionConcurrency is a good example, because it's a parameter we're trying to get rid of in the future by auto-tuning it.

3. Setting the write queue size to `runtime.GOMAXPROCS(0)`, is currently arbitrary. I wanted to set it to a value greater than 0, to enable parallelism in the metamorphic tests. Allowing the metamorphic tests to control the size of the queue isn't useful, because they're just controlling the number of elements added to the queue before it blocks. There is no additional concurrency, or edge cases being tested.

4. In a future pr, I plan to set the writeQueue size such that the amount of disk IO done is comparable to the current master.
</blockquote></details>

The goal to have fewer tunable parameters is correct, but that's a user-facing goal, and the question then becomes "where in the system do we decide what the parameters are?". From that I don't think it's fair to say MaxCompactionConcurrency is a parameter we're trying to get rid of, we're just trying to set it higher on some conditions while not necessarily always using the extra concurrency slots.

At any given point of time, we could have multiple sstable writers open under one Cockroach instance: number of stores times (one for each flush happening at once (max 1) + one for each compaction happening at once (max MaxCompactionConcurrency, defaults to 3)) + one for each import happening (no max, but probably only one import at a time) + one for each backup happening + one for each snapshot rebalance happening (arbitrary max, but usually gated by max snapshot rebalance/recovery rates).

What I'm getting at is that the best place to make this determination is likely going to be in Cockroach, as that's where we know the number of concurrent stores as a starting point, and where we have primitive functions to get number of CPUs. Having each of those writers spin up NumCpu number of threads is much slower than having each of them spin up 1 or 2 writer threads, so we do need a middle ground unless we are making each writer aware of how many other writers are open right now. Now maybe the writer is the best place to put that logic down the line, but this PR is just enabling parallelism, it's not adding that oracle in the writer and we should probably defer that conversation until we have a design in mind for the CPU oracle. (If my understanding is out of sync with your working design of the eventual CPU oracle, feel free to correct me).

In short, the question right now (seeing as we are adding an option which goes on disk and is therefore a harder-to-update change) is "should this be a bool or an int" and I'd say a bool is very broad to net a performance benefit in the medium term, because we will need to have different concurrency parameters for multiple stores and the like. I agree that `runtime.GOMAXPROCS(0)` is arbitrary and will probably be changed to something else, but that still won't be tunable from Cockroach in the multi-store cases without an int option. What do you think?

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581)* status: all files reviewed, 5 unresolved discussions (waiting on @itsbilal, @jbowens, @mufeez-amjad, and @nicktrav)

___
*[options.go, line 911 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyhHyEB94yYjZcs-NSi:-MzasbwOAbuBLSY3j8Fj:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/options.go#L911)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This could also just be `writer_parallelism`
</blockquote></details>

Done.
___
*[options.go, line 1129 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyhI0VI3FPY9d83R_cj:-MzasbUM3y7EAmf8kj2D:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/options.go#L1129)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Same
</blockquote></details>

Done.
___
*[sstable/writer.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1581#-My9PkCRCHTeghDTv5wA:-MzatY9965Cy2wGyHaMX:b5iu15s) ([raw file](https://github.com/cockroachdb/pebble/blob/ae80cfc8a413917bc3615afea9645f8e3e45905d/sstable/writer.go#L225)):*
Took a while to get back to this pr, sorry.

> Having each of those writers spin up NumCpu number of threads is much slower than having each of them spin up 1 or 2 writer threads, so we do need a middle ground unless we are making each writer aware of how many other writers are open right now. 

This isn't happening. Each Writer will only ever start one go routine. The size of the write queue is just the number of blocks which can be added to the writeQueue without blocking(assuming that we're adding blocks faster than the queue can process).

We could discuss whether the number of goroutines launched by the compression queue should be a tuneable number since that actually does launch goroutines.

If you want I can still make the size of the writeQueue tuneable, but that setting doesn't have much direct influence on cpu utilization.
___
*[sstable/writer.go, line 361 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1581#-MyTkTYh2RvLmUPDtmwS:-MzatSkOCy3uaFIByZHs:b-wdj8z5) ([raw file](https://github.com/cockroachdb/pebble/blob/ec1c55c0aa545e4161de10af89725b2908475650/sstable/writer.go#L361)):*
<details><summary><i>Previously, mufeez-amjad (Mufeez Amjad) wrote…</i></summary><blockquote>

nit: consider making this a function on `sizeEstimate`
</blockquote></details>

Nice, thanks.



_Code quote:_
```Go
ate.numWrittenEntries + i.size.estimate.numInflightEntries
```

**Review by itsbilal — COMMENTED** [review]
> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581)* status: all files reviewed, 3 unresolved discussions (waiting on @jbowens, @mufeez-amjad, and @nicktrav)

___
*[sstable/writer.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1581#-My9PkCRCHTeghDTv5wA:-N-46wsjAJ5_nh1OpzD4:b4ajw7v) ([raw file](https://github.com/cockroachdb/pebble/blob/ae80cfc8a413917bc3615afea9645f8e3e45905d/sstable/writer.go#L225)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Took a while to get back to this pr, sorry.

> Having each of those writers spin up NumCpu number of threads is much slower than having each of them spin up 1 or 2 writer threads, so we do need a middle ground unless we are making each writer aware of how many other writers are open right now. 

This isn't happening. Each Writer will only ever start one go routine. The size of the write queue is just the number of blocks which can be added to the writeQueue without blocking(assuming that we're adding blocks faster than the queue can process).

We could discuss whether the number of goroutines launched by the compression queue should be a tuneable number since that actually does launch goroutines.

If you want I can still make the size of the writeQueue tuneable, but that setting doesn't have much direct influence on cpu utilization.





</blockquote></details>

Ah okay, I see - I just saw `runtime.GOMAXPROCS` going into `writeQueueSize` so I assumed that was setting the number of goroutines. We should probably still have an int option for writer parallelism that would govern the number of compression queue goroutines spun up. `WriteCompressionConcurrency` could be the option name. That could be an int, and the lower level writer.Parallelism bool coudl still be a bool that's set to `opts.WriteCompressionConcurrency > 1`. 

I see that the compression queue isn't written yet; do we plan to have one central one, or having each writer spin up a compression queue? If this decision is yet to be made, I'd strongly support the latter to prevent one writer from starving out others.

Also, we should definitely add a comment here that states that `writeQueueSize` does not govern number of goroutines spun up, just the size of buffered channels.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581)* status: 7 of 8 files reviewed, 3 unresolved discussions (waiting on @itsbilal, @jbowens, @mufeez-amjad, and @nicktrav)

___
*[sstable/writer.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1581#-My9PkCRCHTeghDTv5wA:-N-97zj5292uTYjiyE_S:bvmnkj2) ([raw file](https://github.com/cockroachdb/pebble/blob/ae80cfc8a413917bc3615afea9645f8e3e45905d/sstable/writer.go#L225)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Ah okay, I see - I just saw `runtime.GOMAXPROCS` going into `writeQueueSize` so I assumed that was setting the number of goroutines. We should probably still have an int option for writer parallelism that would govern the number of compression queue goroutines spun up. `WriteCompressionConcurrency` could be the option name. That could be an int, and the lower level writer.Parallelism bool coudl still be a bool that's set to `opts.WriteCompressionConcurrency > 1`. 

I see that the compression queue isn't written yet; do we plan to have one central one, or having each writer spin up a compression queue? If this decision is yet to be made, I'd strongly support the latter to prevent one writer from starving out others.

Also, we should definitely add a comment here that states that `writeQueueSize` does not govern number of goroutines spun up, just the size of buffered channels.
</blockquote></details>

Compression queue: https://github.com/cockroachdb/pebble/pull/1481. Not merging the compression queue now, since it's simple and unnecessary for metamorphic testing purposes.

I've been using a central queue. Starvation may not be a problem since we can fall back and do the compression in the Writer client goroutine rather than the compression goroutine. It's probably worth more discussion, but I think the discussion is unrelated to this pr which is trying to get some metamorphic testing going so that we can gain some confidence regarding correctness.

Added the comment.

**Review by itsbilal — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581)* status: all files reviewed, 4 unresolved discussions (waiting on @bananabrick, @jbowens, @mufeez-amjad, and @nicktrav)

___
*[options.go, line 490 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1581#-N-YtUSc9KOHFFL9y9XC:-N-YtUSc9KOHFFL9y9XD:b-vxq6d3) ([raw file](https://github.com/cockroachdb/pebble/blob/ccf1dc13b47f93d9c9e71b67b83c6c87ac57a889/options.go#L490)):*
> ```Go
> 		// compress data blocks and write datablocks to disk in parallel with the Writer client
> 		// goroutine.
> 		WriterParallelism bool
> ```

I don't want this to be a long bikeshed, but I still think this should be an int option called `MaxWriterConcurrency` or `WriterCompressionConcurrency`, that the eventual compression queue can use as a max. It could default to GOMAXPROCS or adjusted accordingly in Cockroach based on what it knows about cgroups and everything. The WriterOption can still remain a bool that just gets set to `MaxWriterConcurrency > 0`.

Probably good to get the option right in a future-proof way, seeing as it goes in the options file and everything.
___
*[sstable/writer.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1581#-My9PkCRCHTeghDTv5wA:-N-Ysfk2443ySqcrqTMc:b-vs1pr4) ([raw file](https://github.com/cockroachdb/pebble/blob/ae80cfc8a413917bc3615afea9645f8e3e45905d/sstable/writer.go#L225)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Compression queue: https://github.com/cockroachdb/pebble/pull/1481. Not merging the compression queue now, since it's simple and unnecessary for metamorphic testing purposes.

I've been using a central queue. Starvation may not be a problem since we can fall back and do the compression in the Writer client goroutine rather than the compression goroutine. It's probably worth more discussion, but I think the discussion is unrelated to this pr which is trying to get some metamorphic testing going so that we can gain some confidence regarding correctness.

Added the comment.
</blockquote></details>

Looks good.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581)* status: 6 of 9 files reviewed, 4 unresolved discussions (waiting on @itsbilal, @jbowens, @mufeez-amjad, and @nicktrav)

___
*[options.go, line 490 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1581#-N-YtUSc9KOHFFL9y9XC:-N-_PtA854GK4Ka54PGF:b-73m6vw) ([raw file](https://github.com/cockroachdb/pebble/blob/ccf1dc13b47f93d9c9e71b67b83c6c87ac57a889/options.go#L490)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I don't want this to be a long bikeshed, but I still think this should be an int option called `MaxWriterConcurrency` or `WriterCompressionConcurrency`, that the eventual compression queue can use as a max. It could default to GOMAXPROCS or adjusted accordingly in Cockroach based on what it knows about cgroups and everything. The WriterOption can still remain a bool that just gets set to `MaxWriterConcurrency > 0`.

Probably good to get the option right in a future-proof way, seeing as it goes in the options file and everything.
</blockquote></details>

Nice catch!

**Review by itsbilal — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1581#-:-N0X2sbYDc5kYYuxAbjo:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @jbowens, @mufeez-amjad, and @nicktrav)
